### PR TITLE
<tuple>: tuple_cat for tuple-like types

### LINKED ISSUE #260: <tuple>: tuple_cat() should be generalized beyond pair and array
When I overhauled `tuple_cat()` long ago, I special-cased `pair` and `array`:

https://github.com/microsoft/STL/blob/d0d724fb916c59f1c4f2830330ed967a72d2d2b8/stl/inc/tuple#L870-L893

This is unnecessary. We could simply use `tuple_size` and `tuple_element` to accept anything resembling a `tuple`.

The Standardese is N4835 [tuple.creation]/12 "[*Note:* An implementation may support additional types in the template parameter pack `Tuples` that support the `tuple`-like protocol, such as `pair` and `array`. -end note]"

### PR DESCRIPTION
**Pull request by ArtemSarmini**
# Description

Closes #260. I also replaced recursive impl of `_View_as_tuple` for arrays with `index_sequence`.

# Checklist

Be sure you've read README.md and understand the scope of this repo.

If you're unsure about a box, leave it unchecked. A maintainer will help you.

- [x] Identifiers in product code changes are properly `_Ugly` as per
  https://eel.is/c++draft/lex.name#3.1 or there are no product code changes.
- [ ] The STL builds successfully and all tests have passed (must be manually
  verified by an STL maintainer before automated testing is enabled on GitHub,
  leave this unchecked for initial submission).
- [x] These changes introduce no known ABI breaks (adding members, renaming
  members, adding virtual functions, changing whether a type is an aggregate
  or trivially copyable, etc.).
- [x] These changes were written from scratch using only this repository,
  the C++ Working Draft (including any cited standards), other WG21 papers
  (excluding reference implementations outside of proposed standard wording),
  and LWG issues as reference material. If they were derived from a project
  that's already listed in NOTICE.txt, that's fine, but please mention it.
  If they were derived from any other project (including Boost and libc++,
  which are not yet listed in NOTICE.txt), you *must* mention it here,
  so we can determine whether the license is compatible and what else needs
  to be done.

### CODE REVIEW

**Comment by ArtemSarmini** [comment]
Tested with
```c++
#include <tuple>
#include <type_traits>

template <class A, class B, class C>
class triple {
	A a;
	B b;
	C c;
};

namespace std {
	template<class A, class B, class C>
	struct tuple_size<triple<A, B, C>> {
		static constexpr size_t size = 3;
	};

	template <class A, class B, class C>
	struct tuple_element<0, triple<A, B, C>> {
		using type = A;
	};

	template <class A, class B, class C>
	struct tuple_element<1, triple<A, B, C>> {
		using type = B;
	};

	template <class A, class B, class C>
	struct tuple_element<2, triple<A, B, C>> {
		using type = C;
	};
}

using namespace std;

// empty, simple
static_assert(is_same_v<
	decltype(tuple_cat(
		declval<tuple<>>(),
		declval<tuple<>>()
	)),
	tuple<>
>);

// pair
static_assert(is_same_v<
	decltype(tuple_cat(
		declval<pair<int, char>>(),
		declval<pair<char, long>>()
	)),
	tuple<int, char, char, long>
>);

// array
static_assert(is_same_v<
	decltype(tuple_cat(
		declval<tuple<double>>(),
		declval<array<int, 2>>()
	)),
	tuple<double, int, int>
>);

// cv and refs are preserved
static_assert(is_same_v<
	decltype(tuple_cat(
		declval<tuple<int&>>(),
		declval<tuple<void* volatile>>()
	)),
	tuple<int&, void* volatile>
>);

// tuple-like
static_assert(is_same_v<
	decltype(tuple_cat(
		declval<tuple<double>>(),
		declval<triple<const int&, void*, size_t>>()
	)),
	tuple<double, const int&, void*, size_t>
>);

int main() {}
```

File: stl/inc/tuple [hunk 1]
```diff
@@ -866,10 +866,40 @@ _NODISCARD constexpr _Ty&& get(array<_Ty, _Size>&& _Arr) noexcept;
...
 
+// STRUCT TEMPLATE _Repeat_type_in_tuple
+template <class _Ty, class... _Meta>
+struct _Repeat_type_in_tuple {
+    static_assert(_Always_false<_Ty>, "Requires index_sequence as second template argument.");
+};
+
+template <class _Ty, size_t... _Ix>
+struct _Repeat_type_in_tuple<_Ty, index_sequence<_Ix...>> {
+    using type = tuple<typename remove_reference_t<decltype(_Ix, declval<_Identity<_Ty>>())>::type...>;
```
Comment (ArtemSarmini) [inline stl/inc/tuple]: My initial idea was
```c++
    using type = tuple<tuple_element_t<_Ix - _Ix, tuple<_Ty>>>;
```
but I think it requires more instantiations

**Comment by ArtemSarmini** [comment]
Duh. clang-format was killed after an hour. I checked in vs, format is ok. Restart pipelines please?

File: stl/inc/tuple [hunk 2]
```diff
@@ -866,10 +866,40 @@ _NODISCARD constexpr _Ty&& get(array<_Ty, _Size>&& _Arr) noexcept;
...
+};
+
+// STRUCT TEMPLATE _Is_tuple_like
```
Comment (CaseyCarter) [inline stl/inc/tuple]: Why not variable template?

File: stl/inc/tuple [hunk 3]
```diff
@@ -866,10 +866,40 @@ _NODISCARD constexpr _Ty&& get(array<_Ty, _Size>&& _Arr) noexcept;
...
+template <class _Ty, class = int>
+struct _Is_tuple_like : false_type {};
+
+template <class _Ty>
+struct _Is_tuple_like<_Ty, decltype((void)tuple_size<_Ty>::size, 0)> : true_type {};
```
Comment (CaseyCarter) [inline stl/inc/tuple]: This partial specialization will never be matched; did you mean to `class = void` on 881?

Comment (CaseyCarter) [reply stl/inc/tuple]:
```suggestion
struct _Is_tuple_like<_Ty, decltype((void)tuple_size<_Ty>::value, 0)> : true_type {};
```

File: stl/inc/tuple [hunk 4]
```diff
@@ -866,10 +866,40 @@ _NODISCARD constexpr _Ty&& get(array<_Ty, _Size>&& _Arr) noexcept;
 template <size_t _Idx, class _Ty, size_t _Size>
 _NODISCARD constexpr const _Ty&& get(const array<_Ty, _Size>&& _Arr) noexcept;
 
+// STRUCT TEMPLATE _Repeat_type_in_tuple
+template <class _Ty, class... _Meta>
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/tuple]: I think `_Indices` is a better name than `_Meta`. (and on 888.)

File: stl/inc/tuple [hunk 5]
```diff
@@ -882,14 +912,9 @@ struct _View_as_tuple<pair<_Ty1, _Ty2>> { // view a pair as a tuple
...
-struct _View_as_tuple<array<_Ty, 0>, _Types...> { // view an array as a tuple; ends recursion at 0
-    using type = tuple<_Types...>;
-};
-
-template <class _Ty, size_t _Size, class... _Types>
-struct _View_as_tuple<array<_Ty, _Size>, _Types...>
-    : _View_as_tuple<array<_Ty, _Size - 1>, _Ty, _Types...> { // view an array as a tuple; counts down to 0
+template <class _Ty, size_t _Size>
+struct _View_as_tuple<array<_Ty, _Size>> { // view an array as a tuple
+    using type = typename _Repeat_type_in_tuple<_Ty, make_index_sequence<_Size>>::type;
```
Comment (CaseyCarter) [inline stl/inc/tuple]: This is `using type = _Meta_repeat_n_c<_Size, _Ty, tuple>;` with the definition of `_Meta_repeat_n_c` from `<variant>` which we could push up into `<type_traits>`.

File: stl/inc/tuple [hunk 6]
```diff
@@ -866,10 +866,40 @@ _NODISCARD constexpr _Ty&& get(array<_Ty, _Size>&& _Arr) noexcept;
...
+};
+
 // STRUCT TEMPLATE _View_as_tuple
-template <class _Ty, class... _For_array>
-struct _View_as_tuple { // tuple_cat() supports only tuples, pairs, and arrays
-    static_assert(_Always_false<_Ty>, "Unsupported tuple_cat arguments.");
+template <class _Ty>
+struct _View_as_tuple {
+    static_assert(_Is_tuple_like<_Ty>::value, "Unsupported tuple_cat arguments.");
+    using type = typename _View_custom_as_tuple<_Ty, make_index_sequence<tuple_size<_Ty>::size>>::type;
```
Comment (CaseyCarter) [inline stl/inc/tuple]:
```suggestion
    using type = typename _View_custom_as_tuple<_Ty, make_index_sequence<tuple_size_v<_Ty>>>::type;
```

**Comment by StephanTLavavej** [comment]
Looks like #257 is still happening 😿

File: stl/inc/tuple (hunk 4 above)
[the remark this answers is not in the document]
Comment (ArtemSarmini) [reply stl/inc/tuple]: Agree, will change

**Comment by ArtemSarmini** [comment]
This also applies to e.g. `make_from_tuple`. Should I inspect other places? I also don't know if making `get` in such places unconditionally unqualified is an option from standard and implementation perspective.

File: stl/inc/tuple (hunk 5 above)
Comment (ArtemSarmini) [reply stl/inc/tuple]: Done

File: stl/inc/tuple (hunk 3 above)
Comment (ArtemSarmini) [reply stl/inc/tuple]: I missed that. Now `void_t` is used

File: stl/inc/tuple (hunk 2 above)
Comment (ArtemSarmini) [reply stl/inc/tuple]: Done. I tend to forget variable templates are a thing in metaprogramming, old habits.

**Comment by ArtemSarmini** [comment]
No it doesn't, cppreference tricked me. I'll do something with `get` in `tuple_cat`

**Comment by CaseyCarter** [comment]
> This also applies to e.g. `make_from_tuple`. Should I inspect other places?

In the library portion of the Standard, the name `meow` is implicitly fully-qualified `::std::meow` (per [[contents]/3](http://eel.is/c++draft/contents#3)) unless explicitly stated otherwise. Despite this fact, LWG conventionally `std::`-qualifies `std::forward`  and single-argument `std::move`, I assume to increase confusion.

> I also don't know if making `get` in such places unconditionally unqualified is an option from standard and implementation perspective.

No, it is not: `get<0>(std::tuple<::X::foo>{})`, for example, could call a function template `get` defined in namespace `X` next to `class foo`. We can't damage the standard-mandated behavior in the process of implementing #260's extension.

**Comment by CaseyCarter** [comment]
This PR also needs some test coverage for the newly-allowed cases.

**Comment by BillyONeal** [comment]
Hi there, @ArtemSarmini are you still working on this?

**Comment by StephanTLavavej** [comment]
It looks like @ArtemSarmini hasn't had time to look into this for the last couple of years - totally understandable, since the maintainer team was also incredibly busy! :joy_cat: I'd like to ensure that the original contributor gets credit for looking into this issue, even if they've moved on to other things, so I'll pick up this PR (which is now our oldest active PR and the only triple-digit one). I believe the main issues are:

* Test coverage is needed (as the PR was created before we added the test harness, I believe),
* When I filed #260 (from my personal todo list of "stuff that could be improved"), I wasn't thinking too hard about the consequences of enlarging our support for a non-Standard extension - which is something that we generally try to avoid these days. I was mostly thinking about simplifying our existing support for `pair` and `array`.
* There is now a proposal being reviewed by Library Evolution, [P2165R3](https://wg21.link/P2165R3) "Compatibility between `tuple`, `pair` and *tuple-like* objects", that is highly relevant. In the latest revision R3, it has reduced its scope to types with `std::get`. Although this is not part of the Working Draft, I think this is a good reason to limit any rework here to only using `std::get`, instead of trying to accept `get` as a customization point far in advance of the Standard coming up with a design in this area.

Thus I believe that different code changes are needed, and will force-push this PR to start from today's `main`.

**Comment by StephanTLavavej** [comment]
Actually, the Azure Pipelines bot is acting weird here; I will instead create a fresh PR and use Co-authored-by for crediting.

**Pull request closed without merging**